File: ruby/examples.rb
#!/usr/bin/env ruby

require "./fibonacci"

# Get new fibonacci enumerator
f = Fibonacci.fibonacci

# Get first five elements of the enumerator
t = f.take(5)
puts "First five elements of fibonacci enumerator"
puts "[#{t.join(', ')}]"
puts ""

# Get next five elements of the enumerator
t = f.take(10)
# drop removes the first `n` elements from the enumerator
d = t.drop(5)
puts "Next five elements of fibonacci enumerator"
puts "[#{d.join(', ')}]"
puts ""

# Get first five elements of the enumerator multiplied by five
t = f.take(5)
m = t.map { |elem| elem * 5 }
puts "First five elements of fibonacci enumerator multiplied by five"
puts "[#{m.join(', ')}]"
puts ""

# Get next five elements of the enumerator multiplied by five
t = f.take(10)
d = t.drop(5)
m = d.map { |elem| elem * 5 }
puts "Next five elements of fibonacci enumerator multiplied by five"
puts "[#{m.join(', ')}]"
puts ""

# Lazily get even values of fibonacci
# This means that the filtering will only be done when values are requested
l = f.lazy
s = l.select { |elem| elem.even? }

# Get first 5 even numbers
t = s.take(5)
puts "First five even elements of lazy fibonacci enumerator"
puts "[#{t.force.join(', ')}]"
puts ""

# Get next 5 even numbers
t = s.take(10)
d = t.drop(5)
puts "Next five even elements of lazy fibonacci enumerator"
puts "[#{d.force.join(', ')}]"
puts ""

File: python/examples.py
#!/usr/bin/env python

from fibonacci import fibonacci
from itertools import islice


def main():
    """Run examples for the fibonacci generator"""
    # Make a new fibonacci generator
    f = fibonacci()

    # Get the first five values from the generator
    print("First five values of the fibonacci sequence")
    print(list(islice(f, 5)))
    print()

    # Get the next five values from the generator
    print("Next five values of the fibonacci sequence")
    print(list(islice(f, 5)))
    print()

    # Recreate generator to start at beginning again
    f = fibonacci()

    # Multiply sequence by 5
    # This is computed "lazily." This means that the multiplication
    # will only occur when a value is requested from the generator
    m = map(lambda x: x * 5, f)

    print("Get first five values multiplied by 5")
    print(list(islice(m, 5)))
    print()

    print("Get the next five values multiplied by 5")
    print(list(islice(m, 5)))
    print()


    # Make sequence of only even numbers
    # This also occurs lazily
    f = filter(lambda x: x % 2 == 0, fibonacci())

    print("Get first five even values")
    print(list(islice(f, 5)))
    print()


if __name__ == "__main__":
    main()
